flink-core/src/main/java/org/apache/flink/util/concurrent/FutureUtils.java

```diff
@@ -1261,6 +1261,43 @@ public static void assertNoException(CompletableFuture<?> completableFuture) {
         handleUncaughtException(completableFuture, FatalExitExceptionHandler.INSTANCE);
     }
 
+    /**
+     * Checks that the given {@link CompletableFuture} is not completed exceptionally with the
+     * specified class. If the future is completed exceptionally with the specific class, then try
+     * to recover using a given exception handler. If the exception does not match the specified
+     * class, just pass it through to later stages.
+     *
+     * @param completableFuture to assert for a given exception
+     * @param exceptionClass exception class to assert for
+     * @param exceptionHandler to call if the future is completed exceptionally with the specific
+     *     exception
+     * @return completable future, that can recover from a specified exception
+     */
+    public static <T, E extends Throwable> CompletableFuture<T> handleException(
+            CompletableFuture<T> completableFuture,
+            Class<E> exceptionClass,
+            Function<E, T> exceptionHandler) {
```

Comment:
For `T` it might not make a huge difference but for `E` it can make a difference. That's why I would suggest to change the signature to

```
CompletableFuture<? extends T> completableFuture,
Class<E> exceptionClass,
Function<? super E, ? extends T> exceptionHandler
```

Reply:
makes sense 👍